packages/dd-trace/src/exporters/agent/writer.js

```diff
@@ -1,129 +1,74 @@
         'Datadog-Meta-Tracer-Version': tracerVersion,
-        'X-Datadog-Trace-Count': String(count)
+        'X-Datadog-Trace-Count': '0'
       },
-      lookup: this._lookup
+      lookup: fakeWriter._lookup
     }
 
-    this._setHeader(options.headers, 'Datadog-Meta-Lang', platform.name())
-    this._setHeader(options.headers, 'Datadog-Meta-Lang-Version', platform.version())
-    this._setHeader(options.headers, 'Datadog-Meta-Lang-Interpreter', platform.engine())
-
-    if (this._url.protocol === 'unix:') {
-      options.socketPath = this._url.pathname
+    fakeWriter._setHeader(options.headers, 'Datadog-Meta-Lang', platform.name())
+    fakeWriter._setHeader(options.headers, 'Datadog-Meta-Lang-Version', platform.version())
+    fakeWriter._setHeader(options.headers, 'Datadog-Meta-Lang-Interpreter', platform.engine())
+    if (fakeWriter._url.protocol === 'unix:') {
+      options.socketPath = fakeWriter._url.pathname
     } else {
-      options.protocol = this._url.protocol
-      options.hostname = this._url.hostname
-      options.port = this._url.port
+      options.protocol = fakeWriter._url.protocol
+      options.hostname = fakeWriter._url.hostname
+      options.port = fakeWriter._url.port
     }
 
-    log.debug(() => `Request to the agent: ${JSON.stringify(options)}`)
-
-    platform.metrics().increment(`${METRIC_PREFIX}.requests`, true)
-
-    platform.request(Object.assign({ data }, options), (err, res, status) => {
-      if (status) {
-        platform.metrics().increment(`${METRIC_PREFIX}.responses`, true)
-        platform.metrics().increment(`${METRIC_PREFIX}.responses.by.status`, `status:${status}`, true)
-      } else if (err) {
-        platform.metrics().increment(`${METRIC_PREFIX}.errors`, true)
-        platform.metrics().increment(`${METRIC_PREFIX}.errors.by.name`, `name:${err.name}`, true)
-
-        if (err.code) {
-          platform.metrics().increment(`${METRIC_PREFIX}.errors.by.code`, `code:${err.code}`, true)
-        }
+    const payload = { data: [msgpack.encode([[], []])] }
+    platform.request(Object.assign(payload, options), (err) => {
+      if (err) {
+        this._writer = new Writer04(url, prioritySampler, lookup)
+      } else {
+        this._writer = new Writer05(url, prioritySampler, lookup)
       }
-
-      platform.startupLog.startupLog({ agentError: err })
-
-      if (err) return log.error(err)
-
-      log.debug(`Response from the agent: ${res}`)
-
-      try {
-        this._prioritySampler.update(JSON.parse(res).rate_by_service)
-      } catch (e) {
-        log.error(e)
-
-        platform.metrics().increment(`${METRIC_PREFIX}.errors`, true)
-        platform.metrics().increment(`${METRIC_PREFIX}.errors.by.name`, `name:${e.name}`, true)
+      for (const spans of this._appends) {
+        this._writer.append(spans)
+      }
+      if (this._needsFlush) {
+        this._writer.flush()
+      }
+      const writer = this._writer
+      this.append = function (spans) {
```

Comment:
It would be cleaner to hold the "appends" elsewhere to that replacing the function is not needed. Basically a multi-stage writer, similar to a stream, or maybe just a stream if it's fast enough.

Reply:
I'd prefer not to use a stream, due to their inherent overhead.

We can avoid replacing the function by keeping some state variable like `_writerInitialized` and using an if-block with that in the condition, so append just looks like:

```js
append (spans) {
  if (this._writerInitialized) {
    this._writer.append(spans)
  } else {
    this._appends.push(spans)
  }
}
```

Would that be what you're looking for here? We'd be trading a one-time function replacement for a conditional check on every single call, but I'd imagine that wouldn't be a performance sink. What do you think?